### PR TITLE
Add get_params method and sklearn based cross validation test

### LINKED ISSUE #72: inherit `scikit-learn` BaseEstimator object?
We can copy the `BaseEstimator` class from here (with proper attribution): https://github.com/mne-tools/mne-python/blob/master/mne/decoding/base.py

Then it should work with `scikit-learn` cv objects without having a dependency.

### PR DESCRIPTION
**Pull request by daniel-acuna**
It should Close #72. `BaseEstimator` produces automatic `__repr__` method and prints human readeable rep of object

### CODE REVIEW

**Comment by pavanramkumar** [comment]
@daniel-acuna what is `sample_weight=None` in `score` function? 

ideally, `score()` would accept `y`, `yhat`, `ynull` and `method` as parameters, where `method='deviance'` or `method='pseudo_R2'` 

and accordingly returns a list of scores. that way we don't need to pass `X` which can be large, and it wouldn't need to call `predict()`

**Comment by daniel-acuna** [comment]
@pavanramkumar  Unfortunately (or fortunately!), to conform to sklearn `RegressorMixin`, we need to override the score function which follows the conventions listed [here](http://scikit-learn.org/stable/modules/generated/sklearn.base.RegressorMixin.html). You can still call `pseudo_R2` of course. If this is really inappropriate, we could raise a warning when calling `score`

**Comment by pavanramkumar** [comment]
@daniel-acuna If you define a new `score()` function as you have done, doesn't it override the inherited function?

To conform as much as possible, we can pass `X` and `y` and call `predict()`. We would need another parameter called `method`. We can also raise appropriate warnings (or document really clearly) that `sample_weights` is never used.

File: pyglmnet/pyglmnet.py [hunk 1]
```diff
@@ -123,20 +124,6 @@ def __init__(self, distr='poisson', alpha=0.05,
         self.eta = eta
         set_log_level(verbose)
 
-    def __repr__(self):
```
Comment (jasmainak) [inline pyglmnet/pyglmnet.py]: why did you remove this?

File: pyglmnet/pyglmnet.py [hunk 2]
```diff
@@ -5,6 +5,7 @@
 
 import numpy as np
 from scipy.special import expit
+from sklearn.base import BaseEstimator, RegressorMixin
```
Comment (jasmainak) [inline pyglmnet/pyglmnet.py]: okay, but we can't do this :) we should copy the `BaseEstimator` class ...

unless you want to create a dependency. Let's start simple with just the `BaseEstimator` and see what features we miss from `RegressorMixin`? As far as I remember, `BaseEstimator` should already allow you to do cross-validation ...

File: tests/test_pyglmnet.py [hunk 3]
```diff
@@ -73,3 +77,8 @@ def test_multinomial_gradient():
     y_pred = glm.predict(X)
     assert_equal(y_pred.shape, (10, 2, 2))  # n_classes x n_samples x n_classes
     assert grad_beta0[0] != grad_beta0[1]
+
+def test_inheritance():
+    """GLM should inherent from sklearn's BaseEstimator and RegressorMixin"""
+    glm = GLM()
+    assert isinstance(glm, BaseEstimator) and isinstance(glm, RegressorMixin)
```
Comment (jasmainak) [inline tests/test_pyglmnet.py]: can you add a test for cross-validation ? something like:

```
from sklearn.cross_validation import KFold, cross_val_score

glm = GLM()
cv = KFold(n_samples, 5)
cross_val_score(glm)
```

File: pyglmnet/pyglmnet.py (hunk 1 above)
Comment (daniel-acuna) [inline pyglmnet/pyglmnet.py]: `BaseEstimator` has it already (https://github.com/scikit-learn/scikit-learn/blob/51a765a/sklearn/base.py#L274), you like this representation better?

File: pyglmnet/pyglmnet.py (hunk 2 above)
Comment (daniel-acuna) [inline pyglmnet/pyglmnet.py]: oh my bad... i didn't realize that sklearn dependency need to be out... for cross validation you only need a `fit` method and a `score` method if `scores=None`. I'll fit

File: pyglmnet/pyglmnet.py (hunk 1 above)
Comment (jasmainak) [inline pyglmnet/pyglmnet.py]: Ah cool, I didn't realize that. +1 for removing this :)

File: pyglmnet/pyglmnet.py (hunk 2 above)
Comment (jasmainak) [inline pyglmnet/pyglmnet.py]: okay, it would be great if you can add the score method!

Comment (jasmainak) [inline pyglmnet/pyglmnet.py]: if `score` method is all what you need, maybe we don't even need `BaseEstimator` for now. We can simply add the `score` method and see if it works ...

Comment (daniel-acuna) [inline pyglmnet/pyglmnet.py]: Exactly, that's what I will do

**Comment by daniel-acuna** [comment]
I know that this won't pass `flake8` test because code base from `sklearn` doesn't that I copied that doesn't conform to it! any chance we can add an exception to the `flake8` command line

**Comment by pavanramkumar** [comment]
What about running `flake8` only for the main script?
```
flake8 --count pyglmnet/pyglmnet.py
```

**Comment by jasmainak** [comment]
@daniel-acuna it's possible to add exceptions to certain files files / folders. Something like this: https://github.com/mne-tools/mne-python/blob/master/setup.cfg#L31.

The diff looks pretty big. Is this smaller by any chance: https://github.com/mne-tools/mne-python/blob/master/mne/decoding/base.py#L15-175 ? It also conforms to pep8 I think ...

File: pyglmnet/mini_sklearn.py [hunk 4]
```diff
@@ -0,0 +1,888 @@
...
+        rendered = '({0})'.format(', '.join(result))
+
+        if self.return_annotation is not _empty:
+            anno = formatannotation(self.return_annotation)
+            rendered += ' -> {0}'.format(anno)
+
+        return rendered
+
+
+class BaseEstimatorish:
```
Comment (jasmainak) [inline pyglmnet/mini_sklearn.py]: lol ... I don't mind calling it the same thing :)

**Comment by jasmainak** [comment]
before we merge, we should also smooth out the license issues.

**Comment by jasmainak** [comment]
we can have an `externals/` folder where we put code copied from other libraries and also add the corresponding license files in there.

File: tests/test_pyglmnet.py [hunk 5]
```diff
@@ -73,3 +75,21 @@ def test_multinomial_gradient():
...
     assert grad_beta0[0] != grad_beta0[1]
+
+def test_cv():
+    """Simple CV check"""
+    X, y = make_regression()
+    model_mn = GLM(distr='normal', alpha=0.01, reg_lambda=np.array([0.0, 0.1, 0.2]))
+    model_mn.fit(X, y)
+
+
+    def simple_scorer(obj, X, y):
```
Comment (jasmainak) [inline tests/test_pyglmnet.py]:
```sh
$ python -c 'import this' | sed -n 7p
Flat is better than nested.
```
:)

**Comment by daniel-acuna** [comment]
The lines you suggest only add the BaseEstimator functionality. CV needs a
bunch of functionality to replicate models during every iteration. I will
try to put my code on a diet and see how it goes

**Comment by daniel-acuna** [comment]
Should I put function outside then? I was looking for modularity but it
looks ugly i guess

**Comment by jasmainak** [comment]
if you look at [this](https://github.com/mne-tools/mne-python/blob/master/mne/decoding/csp.py#L18), you might be able to figure out what's needed. Probably just a `transformermixin` and `estimatormixin`? I am sure they use it in an example with `cross_val_score`

**Comment by jasmainak** [comment]
> I will try to put my code on a diet and see how it goes

Yes!!!!

**Comment by daniel-acuna** [comment]
OK, no need for `BaseEstimator`, `pyglmnet` is really slim now!

**Comment by pavanramkumar** [comment]
@daniel-acuna Do I understand correctly that the user is left to use cross_val_score from sklearn to do their own cross-validation? Or do we provide a wrapper for users that calls cross_val_score?

This decision is important and will affect API design for cvpyglmnet which should have tools to estimate optimal lambda

**Comment by daniel-acuna** [comment]
@pavanramkumar yes, the user have to call `cross_val_score` themselves. For `cvpyglmnet`, I think `sklearn` provides a nice template to follow in [`ElasticNetCV`](https://github.com/scikit-learn/scikit-learn/blob/5876f885f04f1abbf46a9cdfb5cb1d5a2ff64f98/sklearn/linear_model/coordinate_descent.py#L1355)

**Comment by pavanramkumar** [comment]
ok, great. is this PR still WIP or can it be merged?

**Comment by daniel-acuna** [comment]
@pavanramkumar I think it is ready but I am not sure if you have a proccess in place to accept PR. At the very least, @jasmainak should give the OK. I could merge it myself too after that!

**Comment by pavanramkumar** [comment]
@daniel-acuna i think @jasmainak has a talk today, we can wait for him to have a look, no rush. 

do you want to rename this PR before merging? 

right now it does a few different things -- maybe choose an appropriate name to avoid confusion later:
- `verbose` fixes
- added `cross_val_score` example to test_`pyglmnet.py`
- added `get_params` method

**Comment by daniel-acuna** [comment]
@pavanramkumar I removed the `verbose` fixes as they were not needed anymore. I think its ready to merge.

File: tests/test_pyglmnet.py [hunk 6]
```diff
@@ -73,3 +75,20 @@ def test_multinomial_gradient():
...
+def test_cv():
+    """Simple CV check"""
+    X, y = make_regression()
+    model_mn = GLM(distr='normal', alpha=0.01, reg_lambda=np.array([0.0, 0.1, 0.2]))
+    model_mn.fit(X, y)
+
+    cv = KFold(X.shape[0], 5)
+
+    # check that it returns 5 scores
+    assert_equal(len(cross_val_score(model_mn, X, y, cv=cv, scoring=simple_cv_scorer)), 5)
```
Comment (jasmainak) [inline tests/test_pyglmnet.py]: should we define a `score` function in our `GLM` object so we don't need to define `simple_cv_scorer` in the tests? It's also simpler for the user ...

**Comment by jasmainak** [comment]
coverage is above 90% that's awesome!

File: tests/test_pyglmnet.py (hunk 6 above)
Comment (daniel-acuna) [inline tests/test_pyglmnet.py]: I agree but that `score` function would need to make several decisions about what it considers a *null* model. We should create a different issue for that situation I think

Comment (jasmainak) [inline tests/test_pyglmnet.py]: yes, let us do that (create a separate issue :)

weren't there two different ways to score and one of them didn't need a `null` model?

Comment (jasmainak) [inline tests/test_pyglmnet.py]: we could provide one as default and the other one as a utility function which can be imported from `pyglmnet.scorers` or `pyglmnet.utils` ...

Comment (daniel-acuna) [inline tests/test_pyglmnet.py]: At least my code is piggybacking on pseudo_R2 which needs a null model. I am not an expert in `poisson` models and I don't know what a simple null model would be there. @pavanramkumar needs to make a decision
